clang/test/CIR/CodeGen/complex.c

```diff
@@ -0,0 +1,130 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir -o - %s | FileCheck %s --check-prefixes=C,CHECK
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -x c++ -fclangir -emit-cir -o - %s | FileCheck %s --check-prefixes=CPP,CHECK
+
+#ifdef __cplusplus__
+#define VOID
+#else
+#define VOID void
+#endif
+
+double _Complex c, c2;
+int _Complex ci, ci2;
+
+volatile double _Complex vc, vc2;
+volatile int _Complex vci, vci2;
+
+void list_init(VOID) {
+  double _Complex c1 = {1.0, 2.0};
+  int _Complex c2 = {1, 2};
+}
+
+//          C: cir.func @list_init()
+//        CPP: cir.func @_Z9list_initv()
+//      CHECK:   %[[#REAL:]] = cir.const #cir.fp<1.000000e+00> : !cir.double
+// CHECK-NEXT:   %[[#IMAG:]] = cir.const #cir.fp<2.000000e+00> : !cir.double
+// CHECK-NEXT:   %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !cir.double -> !cir.complex<!cir.double>
+//      CHECK:   %[[#REAL:]] = cir.const #cir.int<1> : !s32i
+// CHECK-NEXT:   %[[#IMAG:]] = cir.const #cir.int<2> : !s32i
+// CHECK-NEXT:   %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !s32i -> !cir.complex<!s32i>
+//      CHECK: }
+
+void imag_literal(VOID) {
+  c = 3.0i;
+  ci = 3i;
```

Comment:
Here in `imag_literal()`
```
    %0 = cir.const #cir.imag<#cir.fp<3.000000e+00> : !cir.double> : !cir.complex<!cir.double>
    %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
    cir.store %0, %1 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```
How do you plan to distinguish if the `cir.store` is storing to real or imag part when lowering this? Relying on the imag attribute sounds brittle. Another reason to add LLVM lowering as part of the work, it'd have popped up this kind of question / design issue. This should be using the same mechanism as elsewhere:

```
    %0 = cir.const #cir.imag<#cir.fp<3.000000e+00> : !cir.double> : !cir.complex<!cir.double>
    %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
    %3 = cir.complex.get_img %1 : !cir.ptr<!cir.complex<!cir.double>> -> !cir.ptr<!cir.double>
    cir.store %0, %3 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```

Please rename `cir.complex.get_img` to `cir.complex.imag_ptr`, `cir.complex.get_real` to `cir.complex.real_ptr`.

Reply:
As I have removed `#cir.imag` the CIRGen scheme under discussion here now becomes:

```
c = 3.0i;
// %0 = cir.const #cir.fp<3.000000e+00> : !cir.double
// %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
// %2 = cir.complex.get_img %1 : !cir.ptr<!cir.complex<!cir.double>> -> !cir.ptr<!cir.double>
// cir.store %0, %2 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```

However this may be a bit hard to implement during CIRGen since this assignment is basically an assignment between two complex value. The upstream CodeGen tracks the real and imaginary parts of a complex value individually so it can generate such code directly. In CIR we may use some canonicalization passes to do this job?

Reply:
Since a lot has changed since, let's ignore my first comment on this sub-thread. For the test on `c = 3.0i`, we now get:

```
//      CHECK: %[[#REAL:]] = cir.const #cir.fp<0.000000e+00> : !cir.double
// CHECK-NEXT: %[[#IMAG:]] = cir.const #cir.fp<3.000000e+00> : !cir.double
// CHECK-NEXT: %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !cir.double -> !cir.complex<!cir.double>
```

I'm assuming that the result of `cir.complex.create` is then stored to the global, right? If so, this seems good enough already. Am I missing something?

> However this may be a bit hard to implement during CIRGen since this assignment is basically an assignment between two complex values. The upstream CodeGen tracks the real and imaginary parts of a complex value individually so it can generate such code directly.

Can you elaborate on what upstream CodeGen does? I'm not sure I follow.

> In CIR we may use some canonicalization passes to do this job?

It's always a good option, not sure for this specific example there's much we could do though (I guess it depends on the reply above).

Reply:
> I'm assuming that the result of cir.complex.create is then stored to the global, right? If so, this seems good enough already.

Yes. The result of `cir.complex.create` is then stored into the global.

> Can you elaborate on what upstream CodeGen does?

Let's say we're trying to CodeGen the following assignment statement:

```cpp
// int _Complex c;
c = 3i;
```

The statement contains two sub-expressions: the lhs is an lvalue that indicates the global variable `c`, and the rhs is a prvalue that represents a value of type `int _Complex`. To CodeGen the assignment statement, the CodeGen module will recursively CodeGen these two sub-expressions and then emit some store instructions to actually do the assignment.

When CodeGen-ing the rhs, the sub-expression CodeGen will yield a pair of LLVM values `std::pair<llvm::Value *, llvm::Value *>` to the statement CodeGen. This is how the upstream CodeGen represent a prvalue of complex type: the first element of the pair is the real part of the complex prvalue, and the second element is the imaginary part of the complex prvalue. When the statement CodeGen gets the pair of values, it then emits two store instructions to store the real and imaginary part into the global variable `c`.

In CIR, we take a different approach for CIRGen-ing prvalues of complex types. Instead of yielding a pair of values, CIRGen just yields a single `mlir::Value` that represents the whole complex value. Thus it would be more natural to successively generate a single store instruction to store the whole complex value into the global variable, rather than storing the real and imaginary parts separately.

Reply:
A side note here: the literal `3i` is a complex value whose real part is implicitly 0. Thus one cannot just CodeGen / CIRGen the storing of the imaginary part; the storing of the implicit real part is also necessary.